docs/architecture/clients/services/implementation.md

```diff
@@ -0,0 +1,200 @@
+
+:::tip Why a separate ApiService for the domain?
+
+The vast majority of components are simply interested in reading data, not in writing it. We could
+implement different interfaces to solve this, but there is quite a bit of baggage that comes along
+with API calls, so we elected to separate these concerns out into a separate class: the
+`[Domain]ApiService`. This class is injected into any component that needs to write information to
+domain models, and the `[Domain]ApiService` ensures that first that information is synced with the
+server, and then it is broadcast locally to the application through the `Domain[Internal]Service`.
+
+:::
+
+## Domain Updates
+
+The Bitwarden product contains multiple clients. At any given time, a user could be logged in to
+multiple clients, viewing data and making updates. This influences the architecture for our client
+services. Namely, there are at least two reasons that a domain model needs to be updated:
+
+1. **Server Update**: The user updates it in the current client
+2. **Cache-Only Update**: The user updates it in a different client, and Live Sync triggers an
+   update on the current client
+
+For each of these scenarios, we will examine how the classes above work to persist and propagate
+changes across clients.
+
+### Server Update: Updated in the current client
+
+```kroki type=plantuml
+@startuml
+abstract class Internal[Domain]Service #aliceblue {
+  + upsert(item: [Domain])
+  + replace(items: [Domain][])
+}
+class [Domain]Service #aliceblue {
+  # subject : BehaviorSubject<[Domain][]>
+  + observable: Observable<[Domain][]>
+  - updateObservables()
+}
+class [Domain]ApiService #aliceblue {
+  + put[Domain]()
+}
+together {
+class "UI Component"
+class "Other UI Component"
+}
+class ApiService {
+  + send()
+}
+
+"UI Component" --> "[Domain]ApiService"
+"[Domain]ApiService" --> ApiService
+"[Domain]ApiService" --> "Internal[Domain]Service"
+"Internal[Domain]Service" --> "[Domain]Service"
+"[Domain]Service" ..> "Other UI Component" : Observable next()
+@enduml
+```
+
+In the case that the domain is updated in the client, the change will begin with the user modifying
+data in a UI component. The UI component has a dependency on `[Domain]ApiService`, which is
+responsible for transmitting that change to two places:
+
+1. To the server, to persist the change and also notify other clients, and
+2. To the `BehaviorSubject` for that domain that is storing the state of that domain for the current
+   client
+
+You can see those responsibilities above, where `ApiService` updates the server and
+`Internal[Domain]Service` updates the `BehaviorSubject`.
+
+:::note
+
+In our current implementation, `Internal[Domain]Service` is implemented by `[Domain]Service`.
+However, that implementation is abstracted from the `[Domain]ApiService` so that there is not a
+direct dependency. This will allow for independent implementations in the future.
```

Comment:
To my knowledge, there are no plans to make Interanl impls independent from public ones in the future.

The progress I envision is to make `Internal[Domain]Service` internal to the module with only observables/api calls made public by a modules `index.ts` barrel file

Reply:
I removed the callout altogether, since my intuition for the reasoning was completely wrong 😄 .  If you feel it requires clarification I can add something back.